state/estimator.go

```diff
@@ -0,0 +1,147 @@
+package state
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/celestiaorg/celestia-app/v3/pkg/user"
+	apptypes "github.com/celestiaorg/celestia-app/v3/x/blob/types"
+)
+
+// gasMultiplier is used to increase gas limit in case if tx has additional options.
+const gasMultiplier = 1.1
+
+var ErrGasPriceExceedsLimit = errors.New("state: estimated gasPrice exceeds max gasPrice")
+
+type estimator struct {
+	estimatorAddress string
+
+	defaultClientConn *grpc.ClientConn
+	estimatorConn     *grpc.ClientConn
+}
+
+func (e *estimator) Start(context.Context) error {
+	if e.estimatorAddress == "" {
+		return nil
+	}
+
+	conn, err := grpc.NewClient(
+		e.estimatorAddress,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithUnaryInterceptor(grpcRetry()),
+	)
+	if err != nil {
+		return err
+	}
+	e.estimatorConn = conn
+	return nil
+}
+
+func (e *estimator) Stop(context.Context) error {
+	if e.estimatorConn != nil {
+		if err := e.estimatorConn.Close(); err != nil {
+			return err
+		}
+	}
+	e.estimatorConn = nil
+	return nil
+}
+
+func (e *estimator) estimate(
+	ctx context.Context,
+	cfg *TxConfig,
+	client *user.TxClient,
+	msg sdktypes.Msg,
+) (float64, uint64, error) {
+	if cfg.GasPrice() != DefaultGasPrice && cfg.GasLimit() != 0 {
+		return cfg.GasPrice(), cfg.GasLimit(), nil
```

Comment:
so if both GasPrice and GasLimit are set in the config, GasPrice sanity check (checking if it exceeds the MaxGasPrice) can be bypassed? Is the idea here that if the user sets gas themselves that they wouldn't set a max gas price, and if they set a max gas price, that means they probably didn't specify gas and wants the node to calculate for them (within some upper bound)?

We should document this behaviour here

Reply:
comments throughout this function would be nice.

Reply:
Done. Not sure about "comments throughout this function", I don't see a good reason for that, added Go-doc only.